### DISCUSSION TITLE
Access to flash to change boot.py

### DISCUSSION THREAD

**Discussion by fmafma** [micropython/micropython#15593] [General]
Hi!

During my multicast tests, I set up an automatic boot sequence, launching the socket and waiting for data.

But now, my RPi Pico W does not response anymore to mp-shell :o(
I don't even see any activity from my machine kernel when I plug it (unless I go into bootloader mode, so the card is not dead).

I would like to avoid to clear averything and start from scratch; is there a way to access the flash and change the boot.py script?

Thanks.

**Comment by scruss** [comment]
[mpremote](https://docs.micropython.org/en/v1.23.0/reference/mpremote.html) can often break out of a stuck boot.py, with a combination of `mpremote reset` and then `mpremote edit :boot.py`

As a last resort, [raspberrypi/picotool](https://github.com/raspberrypi/picotool) can download the flash contents as a UF2 if the Raspberry Pi Pico is in bootloader mode. There are tools to recover the contents of the filesystem from a uf2 so you can get your programs back

**Reply by fmafma** [reply]
mpremote does see any device; as said, even the linux kernel doesn't...

picotool seems a little bit complicated; I will spend less time by reseting everything. BTW, how do I do that? If I just reflash the firmware, the boot.py file is not modified, and still freezes the board...

**Reply by robert-hh** [reply]
There is a flash_nuke.dfu firmware by RPi which just erases the flash. e.g. here https://learn.adafruit.com/getting-started-with-raspberry-pi-pico-circuitpython/circuitpython

**Reply by fmafma** [reply]
Thanks!

**Comment by fmafma** [comment]
I just ran into the problem again!!!

This time, no socket reading, I had a input() statement to get the password of the wifi connection. I hit Ctrl-C, as I forgot something, and it froze again :o( No access anymore.

Is it not possible to use input() in boot.py?

**Reply by robert-hh** [reply]
At the RP2 port (and others) the USB device is enabled following a hard reset or power cycle AFTER boot.py is performed. However it stays enabled when boot.py runs during soft reboot. So you cannot use the USB device and hence input() during boot.py.

**Comment by fmafma** [comment]
So, I need to put it in a module imported by boot.py?

**Reply by shariltumin** [reply]
Is it possible to use main.py instead of boot.py?

**Reply by robert-hh** [reply]
> So, I need to put it in a module imported by boot.py?

No. Because boot.py has to finish before USB is enabled. Importing and running another module by boot.py does not finish boot.py. The earliest state where USB is reliably enabled is in code of main.py.

**Reply by fmafma** [reply]
Got it! Thanks.
